python.md

```diff
@@ -31,35 +31,77 @@ d = {
 }
 ```
 
-## Exception handling
+## [Declaring exceptions](https://stackoverflow.com/questions/1319615/proper-way-to-declare-custom-exceptions-in-modern-python/26938914#26938914)
+### The example
 
-* When raising exceptions, try to include enough information to help debugging: what was expected? What actually happened?
-* When writing controllers which handle a lot of Exceptions, try to encapsulate the logic within one `try..except`
-
-``` python
-try:
-    job = get_my_job(job_ref) # Raises some exception when job_ref isn't correct
-    job.change_it_somehow('weee') # Raises some other exception just because
-    job.delete('never wanted it anyways') # Raises another exception
-except MyJobWasntFoundException as e: # Handle each Exception individually with a descriptive name
-    log.info(e.message) # Potentially log as error if serious
-except JobCantBeAltered as e:
-    log.info(e.message)
-except JobDeleteFailedException as e:
-    log.info(e.message)
+```
+class MyAppValueError(ValueError):
+    def __init__(self, foo, *args):
+        # Special attribute you desire with your Error,
+        # perhaps the value that caused the error?:
+        self.foo = foo
+
+        # Allow users initialize misc. arguments as any other builtin Error
+        super(MyAppValueError, self).__init__(message, foo, *args)
 ```
 
+There's no need to write your own `__str__` or `__repr__`. The builtin ones
+are very nice.
+
+### The example, expanded
+
+Exception classes you declare **should inherit** (either directly or indirectly)
+from the `Exception` class.  If there's more that can be told about the nature
+of the exceptional situation, pick a more specific exception class to inherit
+from, from the hierarchy of [predefined exceptions](https://docs.python.org/2/library/exceptions.html#exception-hierarchy).
+
+**Name** your exceptions [in accordance with
+PEP8](https://www.python.org/dev/peps/pep-0008/#exception-names), i.e. use
+suffixes like `Error` or `Warning` where appropriate.
+
+**Avoid** passing human-readable messages to exception constructors, prefer
+passing problematic variable values instead and naming the exception class
+descriptively.  Rationale: Exceptions are entities belonging to the realm of
+the code, not to that of the users.  Users aren't being shown raw exceptions,
+they're being show error messages which are an interpretation of cought
+exceptions.   There are considerations like localization and formatting of the
+message that need to be taken into account before showing a message to the
+user.  Trying to address these consideration at the point in code at which the
```

Comment:
Feel free to ignore this suggestion: I'd write "Trying to address such considerations"

Reply:
That's too subtle to a non-native 😉